azurelinuxagent/ga/collect_telemetry_events.py

```diff
@@ -0,0 +1,585 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import json
+import os
+import re
+import threading
+import traceback
+from collections import defaultdict
+
+import azurelinuxagent.common.logger as logger
+from azurelinuxagent.common import conf
+from azurelinuxagent.common.event import EVENTS_DIRECTORY, TELEMETRY_LOG_EVENT_ID, \
+    TELEMETRY_LOG_PROVIDER_ID, add_event, WALAEventOperation, add_log_event, get_event_logger, \
+    CollectOrReportEventDebugInfo, EVENT_FILE_REGEX, parse_event
+from azurelinuxagent.common.exception import InvalidExtensionEventError, ServiceStoppedError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+from azurelinuxagent.common.telemetryevent import TelemetryEvent, TelemetryEventParam, \
+    GuestAgentGenericLogsSchema, GuestAgentExtensionEventsSchema
+from azurelinuxagent.ga.exthandlers import HANDLER_NAME_PATTERN, is_extension_telemetry_pipeline_enabled
+from azurelinuxagent.ga.periodic_operation import PeriodicOperation
+
+
+def get_collect_telemetry_events_handler(send_telemetry_events_handler):
+    return CollectTelemetryEventsHandler(send_telemetry_events_handler)
+
+
+# too-few-public-methods<R0903> Disabled: This class is used as an Enum
+class ExtensionEventSchema(object): # pylint: disable=R0903
+    """
+    Class for defining the schema for Extension Events.
+
+    Sample Extension Event Example:
+        {
+           "Version":"1.0.0.23",
+           "Timestamp":"2018-01-02T22:08:12.510696Z"    //(time in UTC (ISO-8601 standard),
+           "TaskName":"TestRun"                         //Open for publishers,
+           "EventLevel":"Critical/Error/Warning/Verbose/Informational/LogAlways",
+           "Message": "Successful test"                //(max 3K, 3072 characters),
+           "EventPid":"1",
+           "EventTid":"2",
+           "OperationId":"Guid (str)"
+        }
+    """
+    Version = "Version"
+    Timestamp = "Timestamp"
+    TaskName = "TaskName"
+    EventLevel = "EventLevel"
+    Message = "Message"
+    EventPid = "EventPid"
+    EventTid = "EventTid"
+    OperationId = "OperationId"
+
+
+class _ProcessExtensionEventsPeriodicOperation(PeriodicOperation):
+    """
+    Periodic operation for collecting and sending extension telemetry events to Wireserver.
```

Comment:
This doesn't really send the events to wireserver, it just enqueues them for the other thread to send them.

Reply:
Modified the comment